### ISSUE TITLE
[RFC] Editor Yoshi Flow – UI Components for Settings Panel

### ISSUE THREAD

**Issue by tuchk4** [wix/yoshi#1773]
# RFC: UI Components for Settings Panel

Most of the ideas are come from [`@wix/tpa-settings`](https://github.com/wix-private/tpa-settings) which is used on [site-search](https://github.com/wix-private/site-search/tree/master/packages/search-results-settings), [challenges-web-ooi-component](https://github.com/wix-private/strategic-products/tree/master/achievements/web-client/challenges-web-ooi-component), [groups](https://github.com/wix-private/strategic-products/tree/master/social-groups/social-groups-ui/packages/social-groups-app) and [etsy](https://github.com/wix-private/wix-labs/tree/master/etsy-client) OOI applications.

# Table of contents

- [Requirements](#requirements)
  - [Components](#components)
    - [SettingsTabLayout](#settingstablayout)
    - [ColorPicker & OpacityColorPicker](#colorpicker--opacitycolorpicker)
    - [FontAndColorPicker](#fontandcolorpicker)
  - [Dependencies](#dependencies)

# Requirements

1. There is should not be the implementation of base components. Such components should be implemented at [wix-base-ui](https://github.com/wix-private/wix-base-ui).
2. There is should be only composites that are built on top of [wix-base-ui](https://github.com/wix-private/wix-base-ui) lib to improve DX and dev velocity.
3. Only for Settings Panel usage
4. Should directly support types for `value` and `onChange` props as it is defined in [RFC: Styles at Settings Panel](https://github.com/wix/yoshi/issues/1770). So UI components should have Styles Context in dependencies. This is another one reason, why we should implement contexts as separated packages.

### SettingsTabLayout

[SettingsTabLayout at `@wix/tpa-settings`](https://github.com/wix-private/tpa-settings/blob/master/COMPONENTS.md#settings-panel-layouts)

`SettingsTabLayout` is a wrapper on top of `VerticalTabs` from [wix-base-ui](https://github.com/wix-private/wix-base-ui). Provide better DX and features to set help article ids for tabs.

<p align="center">
  <img title="SettingsPanelLayout" src="https://user-images.githubusercontent.com/5034399/69051430-f97fa980-0a0c-11ea-9f9e-0b3bca777ac3.png" width="400px" />
</p>

```js
const SettingsPanel = () => {
  return (
    <SettingsTabLayout>
      <SettingsTabLayout.Tab
        title="Main"
        articleId="xxx-xxx-xxx-xxx"
        Component={() => <SectionHeader>Title</SectionHeader>}
      />
      <SettingsTabLayout.Divider />
      <SettingsTabLayout.Tab
        title="Support"
        articleId="xxx-xxx-xxx-xxx"
        Component={() => (
          <React.Fragment>
            <SectionHeader>Design</SectionHeader>
            <ColorPicker />
          </React.Fragment>
        )}
      />
      <SettingsTabLayout.UpgradeButton title="Upgrade" referrer="my_referrer" />
    </SettingsTabLayout>
  );
};
```

### ColorPicker & OpacityColorPicker

[ColorPicker & OpacityColorPicker at `@wix/tpa-settings`](https://github.com/wix-private/tpa-settings/blob/master/COMPONENTS.md#colorpicker--opacitycolorpicker)

<p align="center">
  <img title="ColorPicker & OpacityColorPicker" src="https://user-images.githubusercontent.com/5140611/68303782-39f63380-00ad-11ea-8aff-32f4eb442f76.png" width="400px" />
</p>

`Color` picker is a wrapper on top of components from [wix-base-ui](https://github.com/wix-private/wix-base-ui).
`ColorPicker` at `wix-base-ui` is just HTML and CSS for the component, and it is not integrated with `Wix.Styles.openColorPicker` and doesn't support site colors palette change event.

Example of [`openColorPicker call at @wix/tpa-settings`](https://github.com/wix-private/tpa-settings/blob/master/src/components/ColorPicker/openColorPicker.ts).

- Should directly support types from Styles Context, described at [RFC: Styles at Settings Panel and Application](https://github.com/wix/yoshi/issues/1770)
- Use Styles Context to implement Site Colors Palette change event.

```js
import { useStyles } from 'rfs-styles';

const DesignTab = () => {
  const styles = useStyles();

  return (
    <ColorPicker
      title={'Color'}
      value={styles.get(myColor)}
      onChange={value => {
        styles.set(myColor, value);
      }}
    />
  );
};
```

### FontAndColorPicker

[FontAndColorPicker at `@wix/tpa-settings`](https://github.com/wix-private/tpa-settings/blob/master/COMPONENTS.md#fontandcolorpicker)

<p align="center">
  <img title="FontAndColorPicker" src="https://user-images.githubusercontent.com/5140611/68311351-1a193c80-00ba-11ea-85ad-cda9f862feac.png" width="400px" />
</p>

`FontAndColorPicker` picker is a composite on top of components from [wix-base-ui](https://github.com/wix-private/wix-base-ui).
It is not integrated with `Wix.Styles.openColorPicker` and `Wix.Styles.openFontPicker` API.

Example of [`openFontPicker call at @wix/tpa-settings`](https://github.com/wix-private/tpa-settings/blob/master/src/components/FontAndColorPicker/BaseFontAndColorPicker.tsx#L98).

- Should directly support types from Styles Context, described at [RFC: Styles at Settings Panel and Application](https://github.com/wix/yoshi/issues/1770)
- Use Styles Context to implement Site Colors Palette change event.

```js
import { useStyles } from 'rfs-styles';

const DesignTab = () => {
  const styles = useStyles();

  return (
    <FontAndColorPicker
      title={'Color & Font'}
      panelTitle={'Font'}
      value={{
        color: styles.get(myColor),
        font: styles.get(myFont),
      }}
      onChange={value => {
        styles.set(myColor, value.color);
        styles.set(myFont, value.font);
      }}
      fontPickerOptions={{
        fontMinSize: 10,
        fontMaxSize: 20,
      }}
    />
  );
};
```

**Comment by ranyitz** [comment]
This is a great idea - got merged to be part of the editor-flow on #2621

**Comment by tuchk4** [comment]
closing this in favor of https://github.com/wix/yoshi/issues/2725
